### DISCUSSION TITLE
ST7920 like LCD

### DISCUSSION THREAD

**Discussion by EliasA97** [olikraus/u8g2#1789] [Q&A]
Hello, I have many problems with my st7920 display using 8080 parallel. On my uno it works fine, but on my mega it shows a bunch of dots everywhere. Is there any display like the st7920 that is more reliable (and faster)? Also excluding the wires needing, what is better? 8080 parallel, software SPI or hardware SPI?

**Comment by olikraus** [comment]
I think SPI is always better (SW SPI or HW SPI is just a U8g2 question). In fact the ST7920 allows SPI also.
From my experience the ST7920 seems to be somehow unreliable. I personally  would prefer displays with a ST75xx controller instead, something like this: https://www.aliexpress.com/item/32800995659.html

However note, that the ST75xx family are 3.3V display controllers and must not be connected directly to Uno or Mega. Either use a 3.3V Arduino board or add a level translation IC (for example 74HC4050) between controller and display. Sometimes the display PCB also contains level translation. I think the display from the link above does not contain any on board level translation and requires a 74HC4050 between Uno/Mega and display.

**Comment by EliasA97** [comment]
Well, I thout that parallel was faster than spi. If I use HW spi, wouldn't it interfere with other spi devices? Do these displays tristate on MISO?

**Reply by olikraus** [reply]
> I thout that parallel was faster than spi.

In theory yes, however this is only true if also the software in the Arduino can transfer one byte in parallel.
As a precondition for this the D0 line of the display has to be connected to bit 0 of a port in the Arduino.
This would put a lot of restrictions on the user. Instead u8g2 allows any assignmend, but as a consequence u8g2 has to split one byte into individal bits and send them via digitalWrite to the pins you have assigned at the constructor. This is very slow.
If you have an Uno, than it will run at 16 Mhz. Let's assume the bit decoding requires 40 clock cycles and the digital write another 100 cycles (it does a lot of checking), then we will have 160kHz only at the parallel port.

Lets assume SPI runs with 2 MHz, then it would transfer 8 bits with 250kHz, which is faster than parallel port.

> Do these displays tristate on MISO?

tristate is only required for multiple bus master, but SPI will have only one master (one sender). Multiple receiver are possible.
And yes, in order to avoid interference with other SPI receivers, each receiver has a so called "chip select" signal. In fact the u8g2 SPI constructor accepts a chip select pin, which is then driven by U8g2 to tell the display that the current SPI data on the SPI bus is intended for the display.

If there is another receiver on the same SPI bus, then this other receiver requires its own "chip select" wire (which is then hopefully driven by the library for that other receiver).
